backend/src/main/java/ai/verta/modeldb/cron_jobs/DeleteEntitiesCron.java

```diff
@@ -0,0 +1,666 @@
+            .append(" ")
+            .append(alias)
+            .append(" WHERE ")
+            .append(alias)
+            .append(".")
+            .append(ModelDBConstants.DELETED)
+            .append(" = :deleted ")
+            .toString();
+
+    Query projectDeleteQuery = session.createQuery(deleteProjectsQueryString);
+    projectDeleteQuery.setParameter("deleted", true);
+    projectDeleteQuery.setMaxResults(this.recordUpdateLimit);
+    LOGGER.debug("Project delete query: {}", projectDeleteQuery.getQueryString());
+    List<ProjectEntity> projectEntities = projectDeleteQuery.list();
+
+    List<String> projectIds = new ArrayList<>();
+    if (!projectEntities.isEmpty()) {
+      List<String> roleBindingNames = new LinkedList<>();
+      for (ProjectEntity projectEntity : projectEntities) {
+        projectIds.add(projectEntity.getId());
+      }
+      // Get roleBindings by accessible projects
+      getRoleBindingsOfAccessibleProjects(projectEntities, roleBindingNames);
+      LOGGER.debug("num bindings after Projects {}", roleBindingNames.size());
+
+      // Remove all role bindings
+      if (!roleBindingNames.isEmpty()) {
+        roleService.deleteRoleBindings(roleBindingNames);
+      }
+
+      Transaction transaction = session.beginTransaction();
+      String updateDeletedStatusExperimentQueryString =
+          new StringBuilder("UPDATE ")
+              .append(ExperimentEntity.class.getSimpleName())
+              .append(" exp ")
+              .append("SET exp.")
+              .append(ModelDBConstants.DELETED)
+              .append(" = :deleted ")
+              .append(" WHERE exp.")
+              .append(ModelDBConstants.PROJECT_ID)
+              .append(" IN (:projectIds)")
+              .toString();
+      Query deletedExperimentQuery = session.createQuery(updateDeletedStatusExperimentQueryString);
+      deletedExperimentQuery.setParameter("deleted", true);
+      deletedExperimentQuery.setParameter("projectIds", projectIds);
+      deletedExperimentQuery.executeUpdate();
+
+      String updateDeletedStatusExperimentRunQueryString =
+          new StringBuilder("UPDATE ")
+              .append(ExperimentRunEntity.class.getSimpleName())
+              .append(" expr ")
+              .append("SET expr.")
+              .append(ModelDBConstants.DELETED)
+              .append(" = :deleted ")
+              .append(" WHERE expr.")
+              .append(ModelDBConstants.PROJECT_ID)
+              .append(" IN (:projectIds)")
+              .toString();
+      Query deletedExperimentRunQuery =
+          session.createQuery(updateDeletedStatusExperimentRunQueryString);
+      deletedExperimentRunQuery.setParameter("deleted", true);
+      deletedExperimentRunQuery.setParameter("projectIds", projectIds);
+      deletedExperimentRunQuery.executeUpdate();
+
+      for (ProjectEntity projectEntity : projectEntities) {
+        session.delete(projectEntity);
+      }
```

Comment:
Instead of issuing N database commands, can we just issue a single delete request? You already have the IDs for what you want to delete.

Reply:
No, we can not do it in a single delete query. if we do like that it will not delete sub entity list (Not support casecade deletion). So i have used `session.delete(projectEntitiy)` for cascade deletion provided by hibernate